source/CIS/cis-5.rst

```diff
@@ -0,0 +1,599 @@
+- The function MUST reject if the signature verification fails for any withdrawal.
+- The function MUST fail if the CCD balance of the ``signer`` is insufficient to do the withdrawal for any withdrawal.
+- A function MUST non-strictly decrease the CCD balance of the ``signer`` public key and non-strictly increase the balance of the ``to`` address or fail for any withdrawal.
+- A withdrawal back to this contract into the ``depositNativeCurrency`` entrypoint MUST be executed as a normal withdrawal.
+- A withdrawal of a CCD amount of zero MUST be executed as a normal withdrawal.
+- A withdrawal of any amount of CCD to a contract address MUST call a CCD receive hook function on the receiving smart contract with a :ref:`ccd receive hook parameter<CIS-5-functions-transfer-ccd-receive-hook-parameter>`.
+- The contract function MUST reject if the CCD receive hook function called on the contract receiving CCDs rejects for any withdrawal.
+- The balance of a public key not owning any CCD amount SHOULD be treated as having a balance of zero.
+- The function MUST transfer the ``serviceFee`` to the ``serviceFeeRecipient`` for every withdrawal if ``serviceFee!=0``.
+
+.. warning::
+
+  Be aware of transferring CCDs to a non-existing account address or contract address.
+  This specification by itself does not include a standard that has to be followed.
+  Checking the existence of an account address/ contract address would ideally be done off-chain before the message is even sent to the smart contract.
+
+.. _CIS-5-functions-withdrawCis2Tokens:
+
+``withdrawCis2Tokens``
+^^^^^^^^^^^^^^^^^^^^^^
+
+The function executes a list of token withdrawals to native accounts and/or smart contracts out of the smart contract wallet.
+This function MUST call the ``transfer`` function on the CIS-2 token contract for every withdrawal.
+
+Parameter
+~~~~~~~~~
+
+The parameter is a list of withdrawals.
+
+It is serialized as: 2 bytes representing the number of withdrawals (``n``) followed by the bytes for this number of withdrawals.
+
+Each withdrawal is serialized as: a :ref:`CIS-5-PublicKeyEd25519` (``signer``), a :ref:`CIS-5-SignatureEd25519` (``signature``),
+a :ref:`CIS-5-TimeStamp` (``expiryTime``), a :ref:`CIS-5-Nonce` (``nonce``), a :ref:`CIS-5-CCDAmount` (``serviceFee``), an :ref:`CIS-5-Address` (``serviceFeeRecipient``),
+the receiving address :ref:`CIS-2-Receiver` (``to``), a :ref:`CIS-5-TokenAmount` (``tokenAmount``), a :ref:`CIS-5-TokenID` (``tokenID``), a :ref:`CIS-5-ContractAddress` (``cis2TokenContractAddress``), and some additional data :ref:`CIS-2-AdditionalData` (``data``)::
+
+  Cis2TokensWithdrawal ::= (signer: PublicKeyEd25519) (signature: SignatureEd25519) (expiryTime: TimeStamp) (nonce: u64) (serviceFee: CCDAmount) (serviceFeeRecipient: Address) (to: Receiver) (tokenAmount: tokenAmount) (tokenId: tokenID) (cis2TokenContractAddress: ContractAddress) (data: AdditionalData)
+
+  Cis2TokensWithdrawParameter ::= (n: Byte²) (withdrawal: Cis2TokensWithdrawalⁿ)
```

Comment:
**Discussion:** Instead of batching the entire message, it makes more sense to me to batch the signed underlying CIS-2 transfer, such that one signature checking can result in multiple transfers from the same account (maybe even use `TransferParameter` defined in CIS-2 in this definition).

Consider the use case where you want to transfer several NFTs, currently each NFT would require a signature check and separate call to `transfer` of the CIS-2 contract.

```suggestion
  Cis2TokensWithdrawal ::= (signer: PublicKeyEd25519) (signature: SignatureEd25519) (expiryTime: TimeStamp) (nonce: u64) (serviceFee: CCDAmount) (serviceFeeRecipient: Address) (cis2TokenContractAddress: ContractAddress) (transferParameter: TransferParameter)

  Cis2TokensWithdrawParameter ::= (withdrawal: Cis2TokensWithdrawalⁿ)
```